### ISSUE TITLE
Pass native database type attributes in the DMMF

### ISSUE THREAD

**Issue by RobertCraigie** [prisma/prisma#10252]
## Problem

Investigating adding [Decimal](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#decimal) support for https://github.com/RobertCraigie/prisma-client-py.

To implement this correctly, I need to know the precision of the field at the database level as the Python [decimal](https://docs.python.org/3/library/decimal.html) module requires this information to work as expected (see the example below).

However, I could not find any native type information in the DMMF.

## Suggested solution

Add a new field to the DMMF

```json
{
  "dmmf": {
    "models": [
      {
        "fields": [
          {
            "nativeAttributes": [
              {
                "name": "Decimal",
                "args": [50, 30]
              }
            ]
          }
        ]
      }
    ]
  }
}
```

## Alternatives

A workaround is to simply set the precision to a very high number, while I do not know what the implications of this would be, I would assume that this solution would massively decrease performance.

It is also not a complete solution, a `Decimal` could still be created using a higher precision than whatever number I would use.

## Additional context

Consider the following situation:

```py
from decimal import Decimal

value = Decimal('1.00000000000000000000000000000000000000000000000000000000000000000000000000009')
print(value + value)
```

outputs:
```
2.000000000000000000000000000
```

This can be fixed, if precision is known / increased:

```py
from decimal import Decimal, getcontext

getcontext().prec = 100
value = Decimal('1.00000000000000000000000000000000000000000000000000000000000000000000000000009')
print(value + value)
```

outputs:
```
2.00000000000000000000000000000000000000000000000000000000000000000000000000018
```

**Comment by RobertCraigie** [comment]
I'm getting a few requests to support the Decimal type in the Python client, would it be possible to implement this soon?

I'm happy to work on this myself  if necessary but I would need a pointer on where to start as I am not familiar with DMMF serialization.

**Comment by halfmatthalfcat** [comment]
I'm also trying to implement a generator for Scala/Slick-based projects and would love access to the underlying types in order to create the correct static Scala types on the client side.

**Comment by gakonst** [comment]
+1 we'd also love this

**Comment by janpio** [comment]
There is a PR at https://github.com/prisma/prisma-engines/pull/2660

Some background why we are not "just" merging this: DMMF is already large for bigger Prisma schemas. Adding native type information could have considerable additional impact on that, while not being used (yet) in our main product, Prisma Client. Modifying the PR to only render the native types for custom generators or on demand (via a manifest property by the generator) might be a way around it, but would add an API or data structure that we would probably not want to use in Prisma Client and hence would _need_ to break at some time in the future (to not maintain 2 parallel implementations). We are not sure yet how to overcome that.

(Posting your concrete use cases would still be super helpful @gakonst)

**Comment by halfmatthalfcat** [comment]
@janpio My specific use case for adding this was to generate stricter Scala types in a custom generator based on the native types specified in the Prisma schema. Right now, we're only confined to the base Prisma types which cause friction when working with deserialized values. For example, Scala has types for UUID (Java), XML, etc. 

It seems like DMMF growth is going to be a problem whether this PR is merged or not. What is the calculus on the Prisma team around scaling this generally?

**Comment by gakonst** [comment]
My use-case is that I have Decimal types in my Prismafile which are supported by the JS bindings, but in order to support them in Prisma-Py we need extra information apparently, c.f.: https://github.com/RobertCraigie/prisma-client-py/issues/106#issuecomment-1064436790

**Comment by janpio** [comment]
Thanks for the use cases, those always help to understand better where you are coming from.

We have many ideas, but no decided upon solution for the DMMF growth. Issues and PRs like these make it clear that we need one. So if you have good and creative suggestions, we would be very open for an issue that suggest an alternative to the current DMMF that would work well for community generators (and thus also for Prisma Client - it's just a generator after all in the end.)

**Comment by utybo** [comment]
Hello, in case it's needed, here's another use case (that is similar to halfmatthalfcat's). I am currently creating a generator that creates [JetBrains Exposed](https://github.com/jetbrains/exposed) classes so that I can use Prisma's schemas and migration system in Kotlin. Since Exposed itself is a full ORM and has its own representation of the various flavors SQL data types, I need to know the exact type of fields in the Prisma schema. This is to ensure that Prisma's and Exposed's internal representation of the database match as close as possible.

This is especially needed for Prisma's `DateTime`, which can be an Exposed `timestamp()`, `date()`, `time()`, or `datetime()` column, which are all represented *very* differently with Exposed and with Kotlin/Java in general (`Instant`, `LocalDateTime`, `LocalTime`, `LocalDate`...). I could better determine the type I need to use if Prisma includes the native DB type in its DMMF.

I need this in order to do the Prisma-to-Exposed type conversion properly. Otherwise, I'll be resorting to clumsily parsing the schema file on my end to try and find the `@db` attributes, which seems to be a horrendous solution :/

**Comment by botika** [comment]
Make two npm libraries, one with the dmmf with the native types and one without it. And that's all. We don't need it always, use features

**Comment by janpio** [comment]
I closed https://github.com/prisma/prisma/issues/6844 as a duplicate of this, it has 9 👍 .
I also closed https://github.com/prisma/prisma/issues/10252 as it also asks for the same change.

**Comment by hariria** [comment]
Hi @janpio have there been any updates on this?

**Comment by janpio** [comment]
No. This is being tracked in an epic at https://github.com/prisma/prisma/issues/18561, but we have not implemented any of this.

**Comment by hariria** [comment]
Thanks for the clarification @janpio. Would be great to see this happen! 

For context, I made a generator that's complete with the exception of this crucial piece that's preventing me from releasing it. I had assumed the db native type annotations would've been there before I started developing so was a bit surprised to find that it wasn't there. 

Anyways, hope we can prioritize this soon, I think the community would really benefit from it!

**Comment by AlexanderArvidsson** [comment]
To anyone coming across this issue, and while we wait for official support from Prisma, I have had great success parsing the Schema manually in my generator via [@mrleebo/prisma-ast](https://github.com/MrLeebo/prisma-ast)!

Here's how I'm using it:
```typescript
generatorHandler({
  onGenerate: async (options: GeneratorOptions) => {
    const builder = createPrismaSchemaBuilder(options.datamodel)
    const models = options.dmmf.datamodel.models

    models.forEach(model => {
      const schemaModel = builder.findByType('model', { name: model.name })

      if (!schemaModel) throw new Error(`Model ${model.name} not found in schema`)

      model.fields.forEach(field => {
        const schemaField = builder.findByType('field', {
          name: field.name,
          within: schemaModel.properties
        })

        if (!schemaField) throw new Error(`Model ${model.name} not found in schema`)

        // schemaField now contains all attributes, if any!
        const dbAttribute = schemaField.attributes?.find(attr => attr.group === 'db')
        if (dbAttribute?.name === "Uuid") {
          // Do something if the field is UUID
        }
      })
    })
  }
})

```

Big disclaimer, since this is essentially AST traversal, the performance would probably not be that great.
Perhaps traversing the whole AST and creating a lookup map for model fields would be faster, so as to not traverse potentially the whole tree for every single field.

Another option would be to ditch DMMF completely and only rely on the AST parser for your generator, but my strategy right now is to use DMMF as the source of truth, then AST fields are an optional enhancement to the output. If the AST for some reason cannot find the field, it will simply not enhance the output with the attribute information and instead log a warning.

**Comment by littleski** [comment]
This would be amazing to have for custom generators :( can't understand why it's not there yet. We cannot unleash the full power of prisma over all languages.

**Comment by pasha-vuiko** [comment]
I have created the PRs to add the functionality:
https://github.com/prisma/prisma-engines/pull/5009
https://github.com/prisma/prisma/pull/25324

I would appreciate your support 🙂
